jwtclient/transport.go

```diff
@@ -0,0 +1,217 @@
+package jwtclient
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+)
+
+// JWTTransport is a custom http.Client Transport which implements RoundTripper and adds JWT to headers.
+type JWTTransport struct {
+	sync.Mutex
+	token                      string
+	expire                     time.Time
+	uid, secret, loginEndpoint string
+	transport                  http.RoundTripper
+}
+
+// NewJWTTransport is a function returns a new instance of JWTTransport.
+func NewJWTTransport(transport http.RoundTripper, opts ...Option) (*JWTTransport, error) {
+	if transport == nil {
+		transport = http.DefaultTransport
+	}
+
+	t := &JWTTransport{
+		transport: transport,
+	}
+
+	for _, opt := range opts {
+		if opt == nil {
+			continue
+		}
+
+		if err := opt(t); err != nil {
+			return nil, err
+		}
+	}
+
+	if t.expire.IsZero() {
+		t.expire = time.Now().Add(time.Hour * 24)
+	}
+
+	if err := t.GenerateJWT(); err != nil {
+		return nil, err
+	}
+
+	return t, nil
+}
+
+// GenerateJWT is a function that generates JWT and makes a POST request to bouncer to sign it.
+func (t *JWTTransport) GenerateJWT() error {
+	t.Lock()
+	defer t.Unlock()
+
+	// TODO: this is very broken with the latest `jwt-go` lib version 3.0.0
+	token := jwt.New(jwt.SigningMethodRS256)
+	token.Claims["uid"] = t.uid
+	token.Claims["exp"] = t.expire.Unix()
+
+	tokenStr, err := token.SignedString([]byte(t.secret))
+	if err != nil {
+		return err
+	}
+
+	if tokenStr == "" {
+		return errors.New("Empty token")
+	}
+
+	authReq := struct {
+		UID   string `json:"uid"`
+		Token string `json:"token,omitempty"`
+	}{
+		UID:   t.uid,
+		Token: tokenStr,
+	}
+
+	b, err := json.Marshal(authReq)
+	if err != nil {
+		return err
+	}
+
+	authBody := bytes.NewBuffer(b)
+	req, err := http.NewRequest("POST", t.loginEndpoint, authBody)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("Content-type", "application/json")
+	client := &http.Client{}
```

Comment:
would rather we cache the client for reuse across calls to this func. also, since we're decorating an underlying transport already .. shouldn't we use that transport for the http client used in this func?